src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                                         rpc::GetAllJobInfoReply *reply,
                                         rpc::SendReplyCallback send_reply_callback) {
   RAY_LOG(INFO) << "Getting all job info.";
-  auto on_done = [reply, send_reply_callback](
+  auto on_done = [this, reply, send_reply_callback](
                      const absl::flat_hash_map<JobID, JobTableData> &result) {
     for (auto &data : result) {
       reply->add_job_info_list()->CopyFrom(data.second);
     }
-    RAY_LOG(INFO) << "Finished getting all job info.";
-    GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+
+    std::shared_ptr<std::atomic<int>> num_processed_jobs =
+        std::make_shared<std::atomic<int>>(0);
+
+    // For each job with a submission id, we query the internal KV store to get the Ray
+    // Job API JobInfo, and send the gRPC reply in the KV callback. If there are *no* jobs
+    // with a submission id, we must send the gRPC reply in the current function instead.
+    // Use this flag to track it.
+    bool any_jobs_with_submission_id = false;
+
+    for (int i = 0; i < reply->job_info_list_size(); i++) {
+      const auto &metadata = reply->job_info_list(i).config().metadata();
+      if (metadata.find("job_submission_id") == metadata.end()) {
```

Comment:
nit: consider saving `find` result to a local iterator to save another lookup at `at` later.

Reply:
Done!